astropy/io/votable/tree.py

```diff
@@ -2527,6 +2529,20 @@ def fields(self):
         """
         return self._fields
 
+    @property
+    def all_fields(self):
+        """
+        A list of :class:`Field` objects describing the types of each
+        of the data columns. Contrary to ``fields``, this property should
+        list every field that's available on disk, included deselected columns.
+        """
+        # last minute update
```

Comment:
Can you please elaborate more here on why this "last minute update" is needed? Wouldn't "all fields" already be defined when a table is read in?

Reply:
I should have written a better comment; I don't think I understand it better than you do now that 3 months have passed. All I can say for sure now is that removing this makes tests fail, so I know it's still needed. It may have been a band-aid from the start: I don't understand how `self._fields` gets out-of-sync with `self._all_fields`, but it does.
I should also note that right now, I am sick, so maybe I'll get a better shot at explaining it in a couple days, but honestly I'm not sure.

Reply:
@tomdonaldson , do you know why? Are you okay with just merging as-is still? Hope you can advise. Thanks!

Reply:
Thanks for noticing this.  I should have asked the same question.  

As @neutrinoceros mentioned, this is working around the cases where `self._fields` gets out of sync with `self._all_fields`.  I think the workaround is a little messy, so would be in favor of cleaning it up unless that proves even messier. 

I found 2 places that get these out of sync.  That's probably all of them, but it's worth checking for more if we clean this up:
- `TableElement.add_field()` (new method) only appends to `self.fields` (thus `self._fields`) but doesn't touch `all_fields`.  `add_field()` is only called in `TableElement.from_table()`, but that's definitely called in some tests.
  -  Note that `_register_field()`, which replaces the old `_add_field()`, does update `_all_fields`.
- Some tests call `table.fields.extend()` which would evade any updating of `all_fields` even if `add_fields()` were changed to handle that.

So the workaround may not be necessary if those tests were modified to call `add_field()` and `add_field()` were updated to keep `_all_fields` in sync.  Since I may be missing some case and because we can't prevent people from updating the `fields` list (HomogeneousList) without making it a sequence or other slightly scary change, one could argue that the workaround is actually the safest approach.  I think I would prefer an attempt to remove the workaround and revisit if the details get too messy.

Reply:
> Some tests call table.fields.extend() which would evade any updating of all_fields even if add_fields() were changed to handle that.

thank you so much for looking into it ! Now that you mention it, I *think* this was my initial motivation for this "last minute update" pattern: if we have tests that do this, my interpretation is that this is something we've supported thusfar and it shouldn't break without a deprecation period. Considering, my approach seems a lot simpler than what it would take to deprecate mutating a list attribute. I would propose to resolve this by extending my comment to include this explanation, what do you think ?

For completeness, I also confirm that the suggested change allows all tests to pass

<details><summary>patch</summary>

```patch
diff --git a/astropy/io/votable/tests/test_converter.py b/astropy/io/votable/tests/test_converter.py
index 35b1cd180c..9cdb7e5651 100644
--- a/astropy/io/votable/tests/test_converter.py
+++ b/astropy/io/votable/tests/test_converter.py
@@ -295,17 +295,15 @@ def test_vararray():
     table = tree.TableElement(votable)
     resource.tables.append(table)
 
-    tabarr = []
     heads = ["headA", "headB", "headC"]
     types = ["char", "double", "int"]
 
     vals = [["A", 1.0, 2], ["B", 2.0, 3], ["C", 3.0, 4]]
     for i in range(len(heads)):
-        tabarr.append(
+        table.add_field(
             tree.Field(votable, name=heads[i], datatype=types[i], arraysize="*")
         )
 
-    table.fields.extend(tabarr)
     table.create_arrays(len(vals))
     for i in range(len(vals)):
         values = tuple(vals[i])
diff --git a/astropy/io/votable/tests/test_vo.py b/astropy/io/votable/tests/test_vo.py
index 3057ede310..37187aa97b 100644
--- a/astropy/io/votable/tests/test_vo.py
+++ b/astropy/io/votable/tests/test_vo.py
@@ -812,12 +812,8 @@ def table_from_scratch():
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            Field(votable, ID="filename", datatype="char"),
-            Field(votable, ID="matrix", datatype="double", arraysize="2x2"),
-        ]
-    )
+    table.add_field(Field(votable, ID="filename", datatype="char"))
+    table.add_field(Field(votable, ID="matrix", datatype="double", arraysize="2x2"))
 
     # Now, use those field definitions to create the numpy record arrays, with
     # the given number of rows
@@ -861,15 +857,15 @@ def test_build_from_scratch(tmp_path):
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            tree.Field(
-                votable, ID="filename", name="filename", datatype="char", arraysize="1"
-            ),
-            tree.Field(
-                votable, ID="matrix", name="matrix", datatype="double", arraysize="2x2"
-            ),
-        ]
+    table.add_field(
+        tree.Field(
+            votable, ID="filename", name="filename", datatype="char", arraysize="1"
+        )
+    )
+    table.add_field(
+        tree.Field(
+            votable, ID="matrix", name="matrix", datatype="double", arraysize="2x2"
+        )
     )
 
     # Now, use those field definitions to create the numpy record arrays, with
@@ -983,12 +979,8 @@ def _run_test_from_scratch_example():
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            Field(votable, name="filename", datatype="char", arraysize="*"),
-            Field(votable, name="matrix", datatype="double", arraysize="2x2"),
-        ]
-    )
+    table.add_field(Field(votable, name="filename", datatype="char", arraysize="*"))
+    table.add_field(Field(votable, name="matrix", datatype="double", arraysize="2x2"))
 
     # Now, use those field definitions to create the numpy record arrays, with
     # the given number of rows
diff --git a/astropy/io/votable/tree.py b/astropy/io/votable/tree.py
index 194ac4d3c3..c31b167f67 100644
--- a/astropy/io/votable/tree.py
+++ b/astropy/io/votable/tree.py
@@ -2536,11 +2536,6 @@ class TableElement(
         of the data columns. Contrary to ``fields``, this property should
         list every field that's available on disk, including deselected columns.
         """
-        # last minute update
-        for field in self._fields:
-            if field not in self._all_fields:
-                self._all_fields.append(field)
-
         return self._all_fields
 
     @property
@@ -2655,6 +2650,7 @@ class TableElement(
 
     def add_field(self, field: Field) -> None:
         self.fields.append(field)
+        self.all_fields.append(field)
 
     def _register_field(self, iterator, tag, data, config, pos) -> None:
         field = Field(self._votable, config=config, pos=pos, **data)
```

</details>

Reply:
(I also pushed a commit with the one line that is neither a change in tests or a removal of my "last minute update")